File: RdFinanceBD/src/ca/mss/rd/mortgage/impl/MortgageAmortizationIteratorFiltered.java
package ca.mss.rd.mortgage.impl;



public class MortgageAmortizationIteratorFiltered extends MortgageAmortizationIteratorByPaymentExtra {
	
	final static public String className = MortgageAmortizationIteratorFiltered.class.getName();
	final static public long serialVersionUID = className.hashCode();

	final private int year;
	final private int term;

	private MortgageAmortizationRow row;

	public MortgageAmortizationIteratorFiltered(MortgageAmortization amortization, int year, int term) {
		super(amortization);
		this.year = year;
		this.term = term;
	}

	/* (non-Javadoc)
	 * @see com.prosperica.mc.impl.MortgageAmortizationIteratorByPayment#hasNext()
	 */
	@Override
	public boolean hasNext() {
		if( super.hasNext() ){
			if( row == null )
				return true;
			else if( year > 0 && row.nextBalanceOutYear > year )
				return false;
			else if( term > 0 && row.nextBalanceTerm > term )
				return false;
			else
				return true;
		} else
			return false;
	}

	/* (non-Javadoc)
	 * @see java.util.Iterator#next()
	 */
	@Override
	public MortgageAmortizationRow next() {
		while( true ){
			row = super.next();
			if( year > 0 && row.balanceOutYear < year )
				continue;
			else if( term > 0  && row.balanceTerm < term )
				continue;
			else
				return row;
		}
	}

}

File: RdFinanceBD/src.cpp/MortgageAmortizationIteratorFiltered.cpp
#include "MortgageAmortizationIteratorFiltered.h"

namespace ca
{
	namespace mss
	{
		namespace rd
		{
			namespace mortgage
			{
				namespace impl
				{

const std::wstring MortgageAmortizationIteratorFiltered::className = MortgageAmortizationIteratorFiltered::typeid::getName();

					MortgageAmortizationIteratorFiltered::MortgageAmortizationIteratorFiltered(MortgageAmortization *amortization, int year, int term) : MortgageAmortizationIteratorByPaymentExtra(amortization), year(year), term(term)
					{
					}

					bool MortgageAmortizationIteratorFiltered::hasNext()
					{
						if (MortgageAmortizationIteratorByPaymentExtra::hasNext())
						{
							if (row == nullptr)
							{
								return true;
							}
							else if (year > 0 && row->nextBalanceOutYear > year)
							{
								return false;
							}
							else if (term > 0 && row->nextBalanceTerm > term)
							{
								return false;
							}
							else
							{
								return true;
							}
						}
						else
						{
							return false;
						}
					}

					ca::mss::rd::mortgage::impl::MortgageAmortizationRow *MortgageAmortizationIteratorFiltered::next()
					{
						while (true)
						{
							row = MortgageAmortizationIteratorByPaymentExtra::next();
							if (year > 0 && row->balanceOutYear < year)
							{
								continue;
							}
							else if (term > 0 && row->balanceTerm < term)
							{
								continue;
							}
							else
							{
								return row;
							}
						}
					}
				}
			}
		}
	}
}
